### PR TITLE
things you could do with mix - part 3

### PR DESCRIPTION
**Pull request by drumusician**

### CODE REVIEW

File: blog/part_3_option_parser.md [hunk 1]
```diff
@@ -0,0 +1,201 @@
+## Things you could do with mix
+
+### Part 3: OptionParser
+
+Hello again! This is part 3 of my blog post exploring things you could do with mix. If you haven't read the first two you can find those and [here](https://www.theguild.nl/things-you-could-do-with-mix/) and [here](https://www.theguild.nl/thing-you-could-do-with-mix-2/). Last time we explored some new DateTime features in Elixir 1.8.0 and changed our output based on when the new year had arrived. The test we wrote for that was very brittle as it relied on the actual time. So let's start by making sure our full test suite is green again.
```
Comment (pascalw) [inline blog/part_3_option_parser.md]:
```suggestion
Hello again! This is part 3 of my blog post series exploring things you could do with mix. If you haven't read the first two you can find those [here](https://www.theguild.nl/things-you-could-do-with-mix/) and [here](https://www.theguild.nl/thing-you-could-do-with-mix-2/). Last time we explored some new DateTime features in Elixir 1.8.0 and changed our output based on when the new year had arrived. The test we wrote for that was very brittle as it relied on the actual time. So let's start by making sure our full test suite is green again.
```

File: blog/part_3_option_parser.md [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+  {:ok, date_time} = NaiveDateTime.new(date, time)
+
+  output = Mix.Tasks.Time.format_output(date_time, "Australia/Sydney")
+  truncated_time = time |> Time.truncate(:second) |> to_string()
+
+  assert output == truncated_time <> " Happy New Year Sydney!"
+end
+```
+
+I have hardcoded the timezone here for now, but that is good enough for now.
```
Comment (pascalw) [inline blog/part_3_option_parser.md]:
```suggestion
I have hardcoded the timezone here, but that is good enough for now.
```

File: blog/part_3_option_parser.md [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+We can make this test pass by creating the function extracting the output from the run function. We'll go ahead and extract the other conversions as private functions as well ending with the implementation below:
```
Comment (pascalw) [inline blog/part_3_option_parser.md]:
```suggestion
We can make this test pass by extracting a few functions that handle formatting the output. We'll go ahead and extract the other conversions as private functions too, ending with the implementation below:
```

File: blog/part_3_option_parser.md [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+defp time(date_time) do
+  date_time
+  |> DateTime.to_naive
+  |> NaiveDateTime.truncate(:second)
+  |> NaiveDateTime.to_time()
+  |> to_string()
+end
+```
+
+Now that was a lot more refactoring than I initially thought I would do, but we pulled this out into separate functions very nicely!
```
Comment (pascalw) [inline blog/part_3_option_parser.md:119]: I agree this is a nice refactoring, but I'm not sure how it's relevant for the subject of this post?

File: blog/part_3_option_parser.md [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+
+As promised in the title of this post, let's talk about [OptionParser](https://hexdocs.pm/elixir/OptionParser.html). Basically `OptionParser` brings us some nice and convenient ways to add aliases and convert our arguments into options and switches.
+
+As always let's test our way into our implementation. We want to be able to add a flag to our clock that will output the time in either 12- or 24-hour format. The current output uses 24 hour time, so we need to convert that to am / pm.
+
+Here is our test. I have created a helpser function to create the am or pm in our test.
```
Comment (pascalw) [inline blog/part_3_option_parser.md]:
```suggestion
Here is our test. I have created a helper function to create the am or pm in our test.
```

File: lib/mix/tasks/time.ex [hunk 6]
```diff
@@ -13,36 +13,80 @@ defmodule Mix.Tasks.Time do
...
+    |> DateTime.to_date()
+    |> Date.day_of_year()
+  end
+
+  defp time(date_time) do
+    date_time
+    |> DateTime.to_naive
+    |> NaiveDateTime.truncate(:second)
+    |> NaiveDateTime.to_time()
+    |> to_string()
```
Comment (pascalw) [inline lib/mix/tasks/time.ex]: If you don't convert this to string you could use `hour`, `minute` and `seconds` fields on the `Time` struct and not have to split the string etc.

File: blog/part_3_option_parser.md (hunk 4 above)
Comment (drumusician) [reply blog/part_3_option_parser.md:119]: yes I agree. Maybe I can make the subject of this post New Years clean up? The OptionParser part is indeed the smallest part of the post.

What do you think?

Comment (drumusician) [reply blog/part_3_option_parser.md:119]: hmm or just name it refactoring, functional style and explain the pattern matching / arity matching of the functions?

**Pull request merged**